### DISCUSSION TITLE
Command ordering?

### DISCUSSION THREAD

**Discussion by tareksander** [gpuweb/gpuweb#4369] [Q&A] [answered]
I don't really understand the implicit synchronization.

Are render passes in each command buffer executed in order?
Are the command buffers in a queue submission executed in order?
Are the submissions executed in order?
And can all draw calls inside a render pass execute in parallel, even if different pipelines are used?
Or does switching pipelines also imply synchronization?

I want to write the stencil buffer before actually drawing, I need to know if I can put the stencil buffer rendering in the same render pass as my actual drawing.

**Comment by Kangz** [comment]
Yes, yes, yes, draw calls are executed in parallel in the same render pass (even between pipelines), however fixed function depth and stencil processing has to happen in order (imagine that there is a queue of "color/stencil/depth" output operations per pixel that are processed in order). So you should be able to render the stencil buffer in a pass and use it in the same pass.

**Reply by tareksander** [reply]
My though was to make a simple pipeline without a fragment shader that just has some triangles and set to always pass the stencil test and increment on passing the stencil test. Then another pipeline for the color rendering. So it's guaranteed when I call draw with the first pipeline that the stencil increments will have finished before the stencil test for the draw with the second pipeline? Or do I have to do something more elaborate?

**Reply by Kangz** [reply, marked as answer]
Yes that ordering on the stencil is guaranteed.

**Reply by tareksander** [reply]
Thanks!
